File: packages/updater-core/android/src/main/java/com/otakit/core/RNManifest.java
package com.otakit.core;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.Map;
import org.json.JSONObject;

/** RN v3 only. Capacitor's existing v2 verifier is unchanged. */
public final class RNManifest {

  public static String canonical(JSONObject m) throws Exception {
    JSONObject signature = m.getJSONObject("signature");
    String encryption = "null";
    if (!m.isNull("encryption")) {
      JSONObject e = m.getJSONObject("encryption");
      encryption =
        string(e, "alg") +
        "|" +
        string(e, "kid") +
        "|" +
        string(e, "wrapNonce") +
        "|" +
        string(e, "wrappedDek") +
        "|" +
        string(e, "nonce");
    }
    return String.join(
      "\n",
      "MANIFEST:3",
      "appId:" + string(m, "appId"),
      "framework:" + string(m, "framework"),
      "platform:" + string(m, "platform"),
      "channel:" + (m.isNull("channel") ? "null" : JSONObject.quote(string(m, "channel"))),
      "version:" + string(m, "version"),
      "sha256:" + string(m, "sha256"),
      "contentHash:" + string(m, "contentHash"),
      "size:" + integer(m, "size"),
      "runtimeVersion:" + string(m, "runtimeVersion"),
      "strategy:" + string(m, "strategy"),
      "forceImmediate:" + m.getBoolean("forceImmediate"),
      "encryption:" + encryption,
      "releaseId:" + string(m, "releaseId"),
      "kid:" + string(signature, "kid"),
      "iat:" + integer(signature, "iat"),
      "exp:" + integer(signature, "exp")
    );
  }

  public static JSONObject verify(
    String json,
    String appId,
    String platform,
    String runtime,
    String channel,
    Map<String, byte[]> keys,
    long now
  ) throws Exception {
    return verify(json, appId, platform, runtime, channel, keys, now, false);
  }

  public static JSONObject verify(
    String json,
    String appId,
    String platform,
    String runtime,
    String channel,
    Map<String, byte[]> keys,
    long now,
    boolean allowExpiredCache
  ) throws Exception {
    if (json.getBytes(StandardCharsets.UTF_8).length > 4 * 1024 * 1024) throw new Exception(
      "Manifest exceeds size limit"
    );
    JSONObject m = new JSONObject(json);
    JSONObject s = m.getJSONObject("signature");
    String actualChannel = m.isNull("channel") ? null : string(m, "channel");
    String strategy = string(m, "strategy");
    if (
      !m.has("channel") ||
      !m.has("encryption") ||
      integer(m, "schemaVersion") != 3 ||
      !"react-native".equals(string(m, "framework")) ||
      !appId.equals(string(m, "appId")) ||
      !platform.equals(string(m, "platform")) ||
      !runtime.equals(string(m, "runtimeVersion")) ||
      !java.util.Objects.equals(channel, actualChannel) ||
      !(m.get("forceImmediate") instanceof Boolean) ||
      integer(m, "size") <= 0 ||
      integer(m, "size") > 100 * 1024 * 1024 ||
      !string(m, "sha256").matches("[a-f0-9]{64}") ||
      !string(m, "contentHash").matches("[a-f0-9]{64}") ||
      string(m, "version").isEmpty() ||
      string(m, "version").length() > 64 ||
      string(m, "version").matches("(?s).*[\\x00-\\x1f\\x7f].*") ||
      (channel != null && !channel.matches("[A-Za-z0-9._-]{1,64}")) ||
      integer(s, "iat") > now + 300 ||
      (!allowExpiredCache && integer(s, "exp") <= now) ||
      integer(s, "exp") <= integer(s, "iat") ||
      (!strategy.equals("zip") && !strategy.equals("deltas"))
    ) throw new Exception("RN manifest target/schema mismatch");
    if (!m.isNull("encryption")) {
      JSONObject e = m.getJSONObject("encryption");
      if (
        !strategy.equals("zip") ||
        !string(e, "alg").equals("AES-256-GCM") ||
        !string(e, "kid").matches("[a-f0-9]{16}") ||
        Base64.getDecoder().decode(string(e, "nonce")).length != 12 ||
        Base64.getDecoder().decode(string(e, "wrapNonce")).length != 12 ||
        Base64.getDecoder().decode(string(e, "wrappedDek")).length != 48
      ) throw new Exception("Invalid RN encryption envelope");
    }
    if (
      strategy.equals("deltas") &&
      (!string(m, "sha256").equals(string(m, "contentHash")) ||
        m.getJSONArray("files").length() < 2 ||
        m.getJSONArray("files").length() > 5000)
    ) throw new Exception("Invalid RN delta identity");
    byte[] key = keys.get(string(s, "kid"));
    if (key == null) throw new Exception("Untrusted RN signing key");
    java.security.interfaces.ECPublicKey publicKey =
      (java.security.interfaces.ECPublicKey) KeyFactory.getInstance("EC").generatePublic(
        new X509EncodedKeySpec(key)
      );
    if (!isP256(publicKey.getParams())) throw new Exception("Expected P-256 signing key");
    Signature verifier = Signature.getInstance("SHA256withECDSA");
    verifier.initVerify(publicKey);
    verifier.update(canonical(m).getBytes(StandardCharsets.UTF_8));
    if (!verifier.verify(Base64.getUrlDecoder().decode(string(s, "sig")))) throw new Exception(
      "Invalid RN manifest signature"
    );
    return m;
  }

  // SEC 2 v2 section 2.4.2, https://www.secg.org/sec2-v2.pdf.
  // Android API 24 supports EC signatures but has no EC AlgorithmParameters provider.
  static boolean isP256(java.security.spec.ECParameterSpec value) {
    java.security.spec.EllipticCurve curve = value.getCurve();
    return (
      curve.getField() instanceof java.security.spec.ECFieldFp &&
      ((java.security.spec.ECFieldFp) curve.getField()).getP().equals(
        hex("FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF")
      ) &&
      curve
        .getA()
        .equals(hex("FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFC")) &&
      curve
        .getB()
        .equals(hex("5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B")) &&
      value
        .getGenerator()
        .getAffineX()
        .equals(hex("6B17D1F2E12C4247F8BCE6E563A440F277037D812DEB33A0F4A13945D898C296")) &&
      value
        .getGenerator()
        .getAffineY()
        .equals(hex("4FE342E2FE1A7F9B8EE7EB4A7C0F9E162BCE33576B315ECECBB6406837BF51F5")) &&
      value
        .getOrder()
        .equals(hex("FFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551")) &&
      value.getCofactor() == 1
    );
  }

  private static java.math.BigInteger hex(String value) {
    return new java.math.BigInteger(value, 16);
  }

  static String string(JSONObject object, String key) throws Exception {
    Object value = object.get(key);
    if (!(value instanceof String)) throw new Exception("Expected string field: " + key);
    return (String) value;
  }

  static long integer(JSONObject object, String key) throws Exception {
    Object value = object.get(key);
    if (!(value instanceof Number)) throw new Exception("Expected numeric field: " + key);
    Number number = (Number) value;
    if (
      number.doubleValue() != number.longValue() ||
      Math.abs(number.doubleValue()) > 9007199254740991d
    ) throw new Exception("Invalid numeric field: " + key);
    return number.longValue();
  }
}

File: packages/updater-core/ios/Sources/RNManifest.swift
import CryptoKit
import Foundation

public struct RNManifest: Codable {
  public struct Encryption: Codable {
    public let alg: String
    public let kid: String
    public let wrapNonce: String
    public let wrappedDek: String
    public let nonce: String
  }
  public struct Signature: Codable {
    public let kid: String
    public let sig: String
    public let iat: Int64
    public let exp: Int64
  }
  public struct File: Codable {
    public let path: String
    public let sha256: String
    public let size: Int64
    public let url: String
  }
  public let schemaVersion: Int
  public let appId: String
  public let framework: String
  public let platform: String
  public let channel: String?
  public let version: String
  public let sha256: String
  public let contentHash: String
  public let size: Int64
  public let runtimeVersion: String
  public let strategy: String
  public let forceImmediate: Bool
  public let encryption: Encryption?
  public let releaseId: String
  public let signature: Signature
  public let url: String?
  public let files: [File]?

  public func canonicalPayload() throws -> Data {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.withoutEscapingSlashes]
    let channelValue =
      channel == nil ? "null" : String(data: try encoder.encode(channel!), encoding: .utf8)!
    let encryptionValue =
      encryption.map { "\($0.alg)|\($0.kid)|\($0.wrapNonce)|\($0.wrappedDek)|\($0.nonce)" }
      ?? "null"
    return Data(
      [
        "MANIFEST:3", "appId:\(appId)", "framework:\(framework)", "platform:\(platform)",
        "channel:\(channelValue)",
        "version:\(version)", "sha256:\(sha256)", "contentHash:\(contentHash)", "size:\(size)",
        "runtimeVersion:\(runtimeVersion)",
        "strategy:\(strategy)", "forceImmediate:\(forceImmediate ? "true" : "false")",
        "encryption:\(encryptionValue)",
        "releaseId:\(releaseId)", "kid:\(signature.kid)", "iat:\(signature.iat)",
        "exp:\(signature.exp)",
      ].joined(separator: "\n").utf8)
  }

  public static func verify(
    data: Data, builtin: Artifact, channel: String?, keys: [String: Data], now: Int64,
    allowExpiredCache: Bool = false
  ) throws -> RNManifest {
    guard data.count <= 4 * 1024 * 1024,
      let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any],
      raw.keys.contains("channel"), raw.keys.contains("encryption")
    else { throw LaunchError.incompatibleArtifact }
    let manifest = try JSONDecoder().decode(RNManifest.self, from: data)
    guard
      manifest.schemaVersion == 3 && manifest.framework == "react-native"
        && manifest.platform == builtin.platform && manifest.appId == builtin.appId
        && manifest.runtimeVersion == builtin.runtimeVersion && manifest.channel == channel
        && manifest.size > 0 && manifest.size <= 100 * 1024 * 1024
        && matches(manifest.sha256, "^[a-f0-9]{64}$")
        && matches(manifest.contentHash, "^[a-f0-9]{64}$") && !manifest.version.isEmpty
        && manifest.version.utf16.count <= 64
        && !manifest.version.unicodeScalars.contains(where: { $0.value < 32 || $0.value == 127 })
        && (channel == nil || matches(channel!, "^[A-Za-z0-9._-]{1,64}$"))
        && manifest.signature.iat <= now + 300
        && (allowExpiredCache || manifest.signature.exp > now)
        && manifest.signature.exp > manifest.signature.iat
        && ["zip", "deltas"].contains(manifest.strategy),
      let key = keys[manifest.signature.kid]
    else { throw LaunchError.incompatibleArtifact }
    if let encryption = manifest.encryption {
      guard manifest.strategy == "zip", encryption.alg == "AES-256-GCM",
        matches(encryption.kid, "^[a-f0-9]{16}$"),
        Data(base64Encoded: encryption.wrapNonce)?.count == 12,
        Data(base64Encoded: encryption.nonce)?.count == 12,
        Data(base64Encoded: encryption.wrappedDek)?.count == 48
      else { throw LaunchError.incompatibleArtifact }
    }
    if manifest.strategy == "deltas" {
      guard manifest.sha256 == manifest.contentHash, let files = manifest.files,
        files.count >= 2 && files.count <= 5000
      else { throw LaunchError.incompatibleArtifact }
    }
    let publicKey = try P256.Signing.PublicKey(derRepresentation: key)
    let signature = try P256.Signing.ECDSASignature(
      derRepresentation: decodeURLBase64(manifest.signature.sig))
    guard publicKey.isValidSignature(signature, for: try manifest.canonicalPayload()) else {
      throw LaunchError.incompatibleArtifact
    }
    return manifest
  }

  private static func matches(_ value: String, _ expression: String) -> Bool {
    value.range(of: expression, options: .regularExpression) != nil
  }
  private static func decodeURLBase64(_ value: String) throws -> Data {
    let base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(
      of: "_", with: "/")
    guard
      let data = Data(
        base64Encoded: base64 + String(repeating: "=", count: (4 - base64.count % 4) % 4))
    else { throw LaunchError.incompatibleArtifact }
    return data
  }
}
